ambry-store/src/main/java/com/github/ambry/store/BlobStore.java

```diff
@@ -727,22 +735,32 @@ public short undelete(MessageInfo info) throws StoreException {
         Offset currentIndexEndOffset = index.getCurrentEndOffset();
         if (!currentIndexEndOffset.equals(indexEndOffsetBeforeCheck)) {
           FileSpan fileSpan = new FileSpan(indexEndOffsetBeforeCheck, currentIndexEndOffset);
-          IndexValue value =
-              index.findKey(info.getStoreKey(), fileSpan, EnumSet.allOf(PersistentIndex.IndexEntryType.class));
+          IndexValue value = index.findKey(info.getStoreKey(), fileSpan,
+              EnumSet.of(PersistentIndex.IndexEntryType.DELETE, PersistentIndex.IndexEntryType.UNDELETE));
           if (value != null) {
-            throw new StoreException("Cannot undelete id " + info.getStoreKey() + " since concurrent operation occurs",
-                StoreErrorCodes.Life_Version_Conflict);
+            if (value.isUndelete() && value.getLifeVersion() == revisedLifeVersion) {
+              // Might get an concurrent undelete from both replication and frontend. This is considered as an
+              // successful operation and the exception will be captured by the catch statement below.
+              throw new StoreException(
```

Comment:
Instead of the logic in AmbryRequests to look up the life version again, could the life version be carried inside of the `StoreException`? You could create an `UndeleteStoreException` that extends `StoreException` and then `AmbryRequests` can handle it with an instanceof check:
```
if (e instanceof UndeleteStoreException) {
  response = Response(e.getErrorCode(), ((UndeleteStoreException) e).getLifeVersion());
}
```
I feel like this would keep the boundary between store logic and request handling logic more clear.

Reply:
this is a good idea, updated.